File: python/get_collection_size.py
from math import floor
from math import ceil


@make(inputs=[0, 1, 2], outputs=[3], handler=Output.STORE)
def get_collection_size(ctx, collection, unit, round):

    size_bytes = float(ctx.callback.getCollectionAVU(collection, "dcat:byteSize", "", "0", "false")["arguments"][2])

    if unit == "B":
        size = size_bytes
    elif unit == "KiB":
        size = size_bytes / 1024
    elif unit == "MiB":
        size = size_bytes / 1024 / 1024
    elif unit == "GiB":
        size = size_bytes / 1024 / 1024 / 1024
    elif unit == "TiB":
        size = size_bytes / 1024 / 1024 / 1024
    else:
        ctx.callback.failmsg(-1, "Invalid input for 'unit'. Options are: B | KiB | MiB | GiB | TiB")

    # Do the rounding
    if round == "none":
        size = size
    elif round == "floor":
        size = floor(size)
    elif round == "ceiling":
        size = ceil(size)
    else:
        ctx.callback.failmsg(-1, "Invalid input for 'round'. Options are: none | floor | ceiling")

    return size

File: native_irods_ruleset/misc/getCollectionSize.r
# This rule will retrieve the total size of a collection (folder) in iRODS from an iCAT AVU
# Call with
#
# irule -F getCollectionSize.r "*collection='/nlmumc/projects/P000000001/'" "*unit='GiB'" "*round='ceiling'"
#
# Rounding options
# *round='none' returns float with decimals
# *round='ceiling' returns integer rounded up 
# *round='floor' returns integer rounded down 


irule_dummy() {
    IRULE_getCollectionSize(*collection, *unit, *round, *result);
    writeLine("stdout", *result);
}

IRULE_getCollectionSize(*collection, *unit, *round, *result) {
    *sizeBytes = "0";

    getCollectionAVU(*collection,"dcat:byteSize",*sizeBytes,"0","false")

    if ( *unit == "B" ) {
        *size = double(*sizeBytes);
    } else if ( *unit == "KiB" ) {
        *size = double(*sizeBytes)/1024;
    } else if ( *unit == "MiB" ) {
        *size = double(*sizeBytes)/1024/1024;
    } else if ( *unit == "GiB" ) {
        *size = double(*sizeBytes)/1024/1024/1024;
    } else if ( *unit == "TiB" ) {
        *size = double(*sizeBytes)/1024/1024/1024/1024;
    } else {
        failmsg(-1, "Invalid input for 'unit'. Options are: B | KiB | MiB | GiB | TiB");
    }

    # Do the rounding
    if ( *round == "none") {
        *size = *size;
    } else if ( *round == "floor") {
        *size = floor(*size);
    } else if ( *round == "ceiling") {
        *size = ceiling(*size);
    } else {
        failmsg(-1, "Invalid input for 'round'. Options are: none | floor | ceiling");
    }
    
    *result = *size;
}

INPUT *collection="",*unit="",*round=""
OUTPUT ruleExecOut